framework/src/utils/MooseCoordTransform.C

```diff
@@ -0,0 +1,276 @@
+{
+  _coord_type = coord_type;
+}
+
+MooseCoordTransform::MooseCoordTransform(const InputParameters & params)
+{
+  //
+  // coordinate system transformation. If we have multiple different coordinate system types in our
+  // problem, we take note of it because that can cause issues if there is a non-Cartesian
+  // destination coordinate system
+  //
+  const auto & mesh = *params.getCheckedPointerParam<MooseMesh *>("mesh");
+  const auto & coord_sys = mesh.getCoordSystem();
+  std::unordered_set<Moose::CoordinateSystemType> coord_types;
+  auto map_it = coord_sys.begin();
+  // It's possible that the mesh is not in a complete state
+  if (map_it == coord_sys.end())
+    setCoordinateSystem(Moose::COORD_XYZ);
+  else
+    setCoordinateSystem(map_it->second);
+  for (; map_it != coord_sys.end(); ++map_it)
+    coord_types.insert(map_it->second);
+
+  _has_different_coord_sys = coord_types.size() > 1;
+
+  //
+  // rotation
+  //
+  const bool has_alpha = params.isParamValid("alpha_rotation");
+  const bool has_beta = params.isParamValid("beta_rotation");
+  const bool has_gamma = params.isParamValid("gamma_rotation");
+  const auto & up_direction = params.get<MooseEnum>("what_is_up");
+
+  const Direction rz_z_axis =
+      (_coord_type == Moose::COORD_RZ)
+          ? Direction(static_cast<unsigned int>(int(params.get<MooseEnum>("rz_coord_axis"))))
+          : Y;
+
+  if (has_alpha || has_beta || has_gamma)
+  {
+    if (up_direction.isValid())
+      mooseError("Cannot simultaneously set rotation angles as well as an up direction");
+
+    const auto alpha = (has_alpha ? params.get<Real>("alpha_rotation") : Real(0));
+    const auto beta = (has_beta ? params.get<Real>("beta_rotation") : Real(0));
+    const auto gamma = (has_gamma ? params.get<Real>("gamma_rotation") : Real(0));
+
+    if (_coord_type == Moose::COORD_RZ)
+      setRotation(alpha, beta, gamma, rz_z_axis);
+    else
+      setRotation(alpha, beta, gamma);
+  }
+  else if (up_direction.isValid())
+    setRotation(Direction(static_cast<unsigned int>(int(up_direction))), rz_z_axis);
+
+  //
+  // Scale
+  //
+  if (params.isParamValid("length_units_per_meter"))
+    setLengthUnitsPerMeter(params.get<Real>("length_units_per_meter"));
+}
+
+Point
+MooseCoordTransform::operator()(const Point & point) const
+{
+  Point ret(point);
+  if (_rotate)
+    ret = (*_rotate) * ret;
+  if (_scale)
+    ret = (*_scale) * ret;
+
+  // If this shows up in profiling we can make _translation a pointer
+  ret += _translation;
+
+  // Finally, coordinate system conversions
+  if (_coord_type == Moose::COORD_XYZ && _destination_coord_type == Moose::COORD_RZ)
+  {
+    Real r_squared = 0;
+    for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+      if (i != _destination_rz_z_axis)
+        r_squared += ret(i) * ret(i);
+
+    const auto r = std::sqrt(r_squared);
+    const auto z = ret(_destination_rz_z_axis);
+    ret = 0;
+    ret(_destination_rz_r_axis) = r;
+    ret(_destination_rz_z_axis) = z;
+  }
```

Comment:
what about the other cases?
_coord_type == Moose::COORD_XYZ && _destination_coord_type == Moose::COORD_RSPH
_coord_type == Moose::COORD_RZ && _destination_coord_type == Moose::COORD_XYZ
etc

Reply:
RSPH will require special treatment. However, if you are RZ and the "destination" is XYZ then we don't do anything because we are going to choose the RZ system as the "canonical" coordinate system. You can collapse the three-dimensionality of XYZ into the two-dimensionality of RZ in a unique way but you cannot do the reverse. This should be reflected in the new doxygen I added but I will also include it in MooseDocs

Reply:
true true. Well I ll look a the moose docs then when you have it. 
There's also 2D XY -> RZ and 2D XZ -> RZ which could be mentioned